### PR TITLE
[ir][refactor] Pass a context object to `Expression::flatten()`

### PR DESCRIPTION
**Pull request by k-ye**
This PR moves one step forward towards removing the global variable `current_block`

https://github.com/taichi-dev/taichi/blob/14094f25004e1052f134885fddd9252e004eb398/taichi/ir/ir.h#L1906-L1908

I'm trying to do some refactoring to make it easier for separating out the frontend stmts/expressions. Knowing the current block could be useful during expression flattening, but it wasn't done properly because of that global variable. As a fix, I added a small `FlattenContext` wrapper, and let the visitors set the current block during AST traversal.

Lots of the changes here are just for adopting the context. The only real change is here

https://github.com/taichi-dev/taichi/blob/14094f25004e1052f134885fddd9252e004eb398/taichi/ir/ir.h#L1923-L1925.

Note that `IdExpression::flatten` no longer needs to depend on the global `current_block`.

If you are wondering why I'm using a pointer rather than a reference, I'm mostly following the Google coding style on [input/output parameters](https://google.github.io/styleguide/cppguide.html#Reference_Arguments). In short, it recommends that reference params must be marked const and immutable. For mutable output params, use a pointer to explicitly indicate its mutability. This is also consistent with how C would do it.

---

Some further cleanups:

* Remove the global `current_block` completely.
* There are already a few visitors showing this pattern of saving the current block on entering, then restoring it at exit. We can probably create an RAII guard here.

Related issue = #689 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
Can't catch the point exactly... why remove the global `current_block`? If we are afraid of MT-safety then we can use `this->current_block`? Why `FlattenContext`?

File: taichi/transforms/lower_ast.cpp [hunk 1]
```diff
@@ -13,17 +17,27 @@ std::vector<T *> make_raw_pointer_list(
...
+}  // namespace
+
 // Lower Expr tree to a bunch of binary/unary(binary/unary) statements
 // Goal: eliminate Expression, Identifiers, and mutable local variables. Make
 // AST SSA.
 class LowerAST : public IRVisitor {
  private:
   Stmt *capturing_loop;
   std::unordered_set<Stmt *> detected_fors_with_break;
+  Block *current_block_;
```
Comment (xumingkuan) [inline taichi/transforms/lower_ast.cpp:29]: Why not just use `current_block`?

File: taichi/transforms/lower_ast.cpp [hunk 2]
```diff
@@ -4,6 +4,10 @@
 
 TLANG_NAMESPACE_BEGIN
 
+namespace {
+
+using FlattenContext = Expression::FlattenContext;
```
Comment (xumingkuan) [inline taichi/transforms/lower_ast.cpp:9]: Is it the same as `using Expression::FlattenContext;` (avoid typing the class name twice)? Or `using xxx = ...` is a better coding style?

File: taichi/transforms/lower_ast.cpp [hunk 3]
```diff
@@ -121,21 +134,21 @@ class LowerAST : public IRVisitor {
...
+    auto cond_stmt = fctx.back_stmt();
 
     auto &&new_while = std::make_unique<WhileStmt>(std::move(stmt->body));
     auto mask = std::make_unique<AllocaStmt>(DataType::i32);
     new_while->mask = mask.get();
     auto &stmts = new_while->body;
-    for (int i = 0; i < (int)flattened.size(); i++) {
-      stmts->insert(std::move(flattened[i]), i);
+    for (int i = 0; i < (int)fctx.stmts.size(); i++) {
+      stmts->insert(std::move(fctx.stmts[i]), i);
```
Comment (xumingkuan) [inline taichi/transforms/lower_ast.cpp:146]: It looks the same as `stmts->insert(std::move(fctx.stmts), 0)`. Although this is probably not directly related to this PR :)

File: taichi/transforms/lower_ast.cpp (hunk 2 above)
Comment (k-ye) [reply taichi/transforms/lower_ast.cpp:9]: Unfortunately, `using Expression::FlattenContext;` simply won't work at the **namespace scope**. Here's the error message:

```
 error: using declaration cannot refer to class member
```

 See https://en.cppreference.com/w/cpp/language/using_declaration

File: taichi/transforms/lower_ast.cpp (hunk 1 above)
Comment (k-ye) [reply taichi/transforms/lower_ast.cpp:29]: Good point, here's my thought:

1. Google's C++ style recommends private class members to end with `_`, see https://google.github.io/styleguide/cppguide.html#Variable_Names
2. We already have a global `current_block`. Having a member variable could be very confusing.

That said, I know that this class didn't follow 1 when it was created, and that we might want to deprecate the global `curretn_block` asap, so I've renamed this to `current_block`. Moreover, whenever I need to refer to it, I've prepended `this->` to distinguish it from the global one.

File: taichi/transforms/lower_ast.cpp (hunk 3 above)
Comment (k-ye) [reply taichi/transforms/lower_ast.cpp]: Nice thanks!

**Comment by k-ye** [comment]
@archibate 

> why remove the global current_block?  If we are afraid of MT-safety then we can use this->current_block? 

TLDR: don't use global variables.

No, it's not MT-safety here that worries me, after all there is no multi-threading involved in this component (at least not before parallel compilation work). Instead, global variable is a philosophical issue, and violates way too many good practices in software engineering. It is one of the easiest ways to pollute the code base.

I'm not trying to set up a doctrine here and say that global variables are always evil. It's like `goto` -- useful, but needs to be used very wisely. 

Global variable should be used when it logically makes sense for that variable to be global. However, if the only justification is "making it global gets my work done", then it reveals a design problem, and it is the design that we need to fix.

If you haven't, I'd suggest search a bit on why global variables are bad. I think any book talking about software engineering will provide a very detailed answer.

**Comment by k-ye** [comment]
> But I also realized that it also bad to make a huge class combined with many unrelated members, or member variables will have no difference as global variables.

Exactly! If you search for [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), you should see that ideally a class should be small, and should be responsible for one thing. I think this not only applies to classes, but functions or any other thing with a scope as well. To conclude, the smaller the scope is, the easier it becomes for us to understand, and therefore change the program. 

> TBH I found Program is some kind of this:

Yeah. I think that's unfortunate, but understandable, given that the work is humongous for a single person to make Taichi work. Now that we do have a few extra people, I think when we change something, we should always aim at helping improve the quality. Otherwise it will be [broken window theory](https://en.wikipedia.org/wiki/Broken_windows_theory) ...

> Thanks for getting rid of the global current_block!

It's still here yet, will do it in another PR..

**Comment by yuanming-hu** [comment]
I'm merging this in since @xumingkuan is taking a couple of days off after our project rush at MIT. This is a very helpful code quality improvement. Thanks!

**Comment by archibate** [comment]
On Thu, Apr 30, 2020 at 09:24:21PM -0700, Yuanming Hu wrote:
> Merged #901 into master.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/taichi-dev/taichi/pull/901#event-3292189903

test
